graph.py

```diff
@@ -0,0 +1,85 @@
+class Graph(object):
+    """
+    Create the graph class and initialize an empty dictionary
+    """
+    def __init__(self):
+        self.g = {}
+
+    def nodes(self):
+        """
+        returns a list of nodes
+        """
+        return self.g.keys()
+
+    def edges(self):
+        """
+        returns a list of tuples the corresponds to the edges
+        """
+        result = []
+        for key, values in self.g.iteritems():
+            for value in values:
+                if value:
+                    result.append((key, value))
+        return result
+
+    def add_node(self, n):
+        """
+        Add a node to the Graph
+        """
+        self.g[n] = []
+
+    def add_edge(self, n1, n2):
+        """
+        Add an edge between two nodes
+
+        If any of the nodes were not created then they will be created
+        """
+        if not self.g.has_key(n1):
+            self.add_node(n1)
+        if not self.g.has_key(n2):
+            self.add_node(n2)
+        self.g[n1].append(n2)
+
+    def del_node(self, n):
+        """
+        Delete an existing node from the Graph
+        """
+        try:
+            del self.g[n]
+        except KeyError:
+            raise IndexError("Node doesn't exist")
```

Comment:
What happens if there are edges leading to this node from elsewhere in the graph?  Write a test that exposes the problem and then solve it.